### PR TITLE
fix bottom message hidden behind composer with minimized keyboard when replying to message

### LINKED ISSUE #20307: Bottom message hidden behind composer with minimised keyboard when replying to message
# Bug Report

## Problem 

After hiding the `Jump to` button, there is no more space between the last message in the story and the composer. This causes the bottom message in the story to be partially hidden behind the composer if it contains a reply and the keyboard is collapsed. We should either increase the scroll area or the padding above the composer (where the `Jump to` button used to be). 

#### Expected behavior
The last message is visible

#### Actual behavior
https://github.com/status-im/status-mobile/assets/67952253/8d28b09e-4bd1-44b3-b2fd-9bcf2319fe10

### Reproduction

1. Open chat with some history
2. Long-press a message
3. Reply
4. Collapse the keyboard
5. Try to scroll up to see that last message

### Additional Information

- Status version: RC 2.29 
- Operating System: Android, iOS

[comment]: # (Please, add logs/notes if necessary)

### PR DESCRIPTION
**Pull request by Parveshdhull**
fixes https://github.com/status-im/status-mobile/issues/20307

### Testing
Please test last message with all, edit, reply, image, link etc.

status: ready

### CODE REVIEW

File: src/status_im/contexts/chat/messenger/messages/list/view.cljs [hunk 1]
```diff
@@ -102,18 +102,37 @@
      ;; https://github.com/status-im/status-mobile/issues/17426
      [quo/skeleton-list (skeleton-list-props :messages parent-height platform/ios?)]]))
 
+(defn header-height
+  [insets able-to-send-message?]
+  (let [images         (rf/sub [:chats/sending-image])
+        reply          (rf/sub [:chats/reply-message])
+        edit           (rf/sub [:chats/edit-message])
+        link-previews? (or (rf/sub [:chats/link-previews?])
+                           (rf/sub [:chats/status-link-previews?]))]
```
Comment (clauxx) [inline src/status_im/contexts/chat/messenger/messages/list/view.cljs]: is it ok to have subscriptions in the body of non-component functions?

Comment (Parveshdhull) [reply src/status_im/contexts/chat/messenger/messages/list/view.cljs]: hi @clauxx, thank you for reviewing PR.

I think these subscriptions are still bound to the component, as they are directly being called inside the component function itself. We use same approach for [composer](https://github.com/status-im/status-mobile/blob/d7530dfbee64aa6e042b63c402682dbe10b95b43/src/status_im/contexts/chat/messenger/composer/utils.cljs#L191). Also to be sure, I checked re-frisk and didn't saw any subscription leaks.

An example of the leak, where we used subscription in the callback function
https://github.com/status-im/status-mobile/pull/15878/files

@flexsurfer probably knows better about this. Please feel free to correct me.

Comment (Parveshdhull) [reply src/status_im/contexts/chat/messenger/messages/list/view.cljs]: tried calling this function in setTimeout, got this
image

Comment (flexsurfer) [reply src/status_im/contexts/chat/messenger/messages/list/view.cljs]: yeah that's fine from re-frame pov, but from the code I would prefer to always have all subscriptions in the component let block

**Review by flexsurfer — APPROVED** [review]

File: src/status_im/contexts/chat/messenger/messages/list/view.cljs (hunk 1 above)
Comment (Parveshdhull) [reply src/status_im/contexts/chat/messenger/messages/list/view.cljs]: thank you, updated

**Review by clauxx — APPROVED** [review]

**Review by OmarBasem — APPROVED** [review]

**Comment by pavloburykh** [comment]
@Parveshdhull thanks for the PR. Please, take a look at the issues. 

### ISSUE 1 Bottom message is still partially hidden behind composer on IOS

**IOS:** 

photo_2024-06-10 15 20 49

**Android:** on Android message still located to close to composer

photo_2024-06-10 15 20 53

**Comment by pavloburykh** [comment]
### ISSUE 2 Space between messages became too large on Android

Reproduced on Android 12, Samsung Galaxy A52


**Actual result:** 

photo_2024-06-10 15 20 57


**Expected result:** 

photo_2024-06-10 15 21 00

**Comment by pavloburykh** [comment]
### ISSUE 3 Wrong opacity of the top edge of composer after editing message

Not sure if it is PR related. Similar issues with opacity are reproducible in nightly, but looks like they have been partially fixed in this PR except current issue. @Parveshdhull I can log separately depending on your decision. 

**Steps:** 

1. Start editing multiline message
2. Quit editing by tapping [x] button
3. Pay attention at composer opacity

**Actual result:** top edge of composer has wrong opacity

photo_2024-06-10 15 20 42


https://github.com/status-im/status-mobile/assets/97245802/805cd89f-fbbc-4a3e-8475-ed8cbdfda9de

**Comment by Parveshdhull** [comment]
hi @pavloburykh, Thank you very much for testing the PR and finding these issues.

**Issue 1:** This issue happened because of the multi-line message. When we closed the composer with the multi-line message, the composer collapsed with increased height for showing two lines (instead of one) and this was causing overlap. So to fix this, we have to increase the margin of the last message from the composer for multi-line messages. I tried to add this margin when the composer was closing with a multi-line message. But this sudden change in the margin of the last message gave a little jumpy transition. So currently I am adding this margin to the last message as soon as composer input has more than 1 line. Please let me know if this looks ok or needs to be improved.

**Issue 2:** It was not reproducible by me, still this one doesn't look related to PR. Please let me know if it is still happening in the new build and not in develop. We can try reverting all commits in PR and see if it gets fixed.

**Issue 3:** Should be fixed now.

**Comment by pavloburykh** [comment]
> hi @pavloburykh, Thank you very much for testing the PR and finding these issues.
> 
> **Issue 1:** This issue happened because of the multi-line message. When we closed the composer with the multi-line message, the composer collapsed with increased height for showing two lines (instead of one) and this was causing overlap. So to fix this, we have to increase the margin of the last message from the composer for multi-line messages. I tried to add this margin when the composer was closing with a multi-line message. But this sudden change in the margin of the last message gave a little jumpy transition. So currently I am adding this margin to the last message as soon as composer input has more than 1 line. Please let me know if this looks ok or needs to be improved.
> 
> **Issue 2:** It was not reproducible by me, still this one doesn't look related to PR. Please let me know if it is still happening in the new build and not in develop. We can try reverting all commits in PR and see if it gets fixed.
> 
> **Issue 3:** Should be fixed now.

Thanx for the fixes @Parveshdhull! Sorry for testing delay, I have missed this comment and noticed it just now. Will test the fixes tomorrow and provide and updated.

**Comment by pavloburykh** [comment]
@Parveshdhull thank you! All 3 issues are fixed. Ready for merge.

**Pull request merged**